include/picongpu/fields/MaxwellSolver/CCK/Derivative.hpp

```diff
@@ -0,0 +1,162 @@
+                 *
+                 * This derivative can only be applied for the E field.
+                 *
+                 * @tparam T_cherenkovFreeDirection direction to remove numerical Cherenkov
+                 *                                  radiation in, 0 = x, 1 = y, 2 = z
+                 * @tparam T_direction direction to take derivative in, 0 = x, 1 = y, 2 = z
+                 */
+                template<uint32_t T_cherenkovFreeDirection, uint32_t T_direction>
+                struct DerivativeFunctor
+                {
+                private:
+                    //! Internally used derivative functor
+                    using InternalDerivativeFunctor
+                        = differentiation::DerivativeFunctor<differentiation::Forward, T_direction>;
+
+                public:
+                    /** Lower margin: we move by 1 along each direction and
+                     *  apply InternalDerivativeFunctor, add those up
+                     */
+                    using LowerMargin = typename pmacc::math::CT::add<
+                        typename pmacc::math::CT::make_Int<simDim, 1>::type,
+                        typename GetLowerMargin<InternalDerivativeFunctor>::type>::type;
+
+                    /** Upper margin: we move by 1 along each direction and
+                     *  effectively apply InternalDerivativeFunctor (for T_direction not
+                     *  literally, but structurally), add those up
+                     */
+                    using UpperMargin = typename pmacc::math::CT::add<
+                        typename pmacc::math::CT::make_Int<simDim, 1>::type,
+                        typename GetUpperMargin<InternalDerivativeFunctor>::type>::type;
+
+                    HDINLINE DerivativeFunctor()
+                    {
+                    }
+
+                    /** Return derivative value at the given point
+                     *
+                     * @tparam T_DataBox data box type with field data
+                     * @param data position in the data box to compute derivative at
+                     */
+                    template<typename T_DataBox>
+                    HDINLINE typename T_DataBox::ValueType operator()(T_DataBox const& data) const
+                    {
+                        constexpr uint32_t freeDir = T_cherenkovFreeDirection;
+                        constexpr uint32_t dir0 = T_direction;
+                        constexpr uint32_t dir1 = (dir0 + 1) % 3;
+                        constexpr uint32_t dir2 = (dir0 + 2) % 3;
+
+                        constexpr float_64 step[3] = {CELL_WIDTH, CELL_HEIGHT, CELL_DEPTH};
+                        constexpr float_64 step2[3]
+                            = {CELL_WIDTH * CELL_WIDTH, CELL_HEIGHT * CELL_HEIGHT, CELL_DEPTH * CELL_DEPTH};
+
+                        // delta: smallest step size
+                        constexpr float_64 delta = step[freeDir];
```

Comment:
I think you can make use of an `if constexpr` here for the cell sizes and remove the `T_cherenkovFreeDirection` template argument.